File: 프로그래머스/2/67257. ［카카오 인턴］ 수식 최대화/［카카오 인턴］ 수식 최대화.java
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;
import java.util.StringTokenizer;

class Solution {
    private static final String[][] priorityOperators = {
        "+-*".split(""),
        "+*-".split(""),
        "-+*".split(""),
        "-*+".split(""),
        "*+-".split(""),
        "*-+".split(""),
    };
    
    private static List<String> getTokens(String expression) {
        List<String> result = new ArrayList<>();
        
        StringTokenizer st = new StringTokenizer(expression, "+*-", true);
        while (st.hasMoreTokens()) {
            result.add(st.nextToken());
        }
        
        return result;
    }
    
    private static String calculate(String numStr1, String numStr2, String op) {
        Long number1 = Long.parseLong(numStr1);
        Long number2 = Long.parseLong(numStr2);
        
        switch (op) {
            case "+": return String.valueOf(number1 + number2);
            case "-": return String.valueOf(number1 - number2);
            case "*": return String.valueOf(number1 * number2);
            default: return "0";
        }
    }
    
    private static long calculate(List<String> tokens, String[] ops) {
        List<String> expr = new ArrayList<>(tokens);
    
        for (String op : ops) {
            Stack<String> stack = new Stack<>();
            
            for (int i = 0; i < expr.size(); i++) {
                String token = expr.get(i);
                
                if (token.equals(op)) {
                    String prev = stack.pop();
                    String next = expr.get(++i);
                    stack.add(calculate(prev, next, op));
                } else {
                    stack.add(token);
                }
            }
            
            expr = new ArrayList<>(stack);
        }

        Long result = Long.parseLong(expr.get(0));
        return Math.abs(result);
    }
    
    public long solution(String expression) {
        long answer = 0;
        List<String> tokens = getTokens(expression);
        
        for (String[] ops : priorityOperators) {
            answer = Math.max(answer, calculate(tokens, ops));
        }
        
        return answer;
    }
}

File: 프로그래머스/2/67257. ［카카오 인턴］ 수식 최대화/［카카오 인턴］ 수식 최대화.py
import re

from collections import deque
from itertools import permutations
from math import inf


def solution(expression):
    answer = -inf
    numbers = list(re.sub('\D', '|', expression).split('|'))
    ops = list(re.sub('\d', '', expression))
    
    exp = [numbers[0]]
    for i, op in enumerate(ops):
        exp.append(op)
        exp.append(numbers[i + 1])
    
    for priority_ops in permutations('+-*', 3):
        result = calculate_result(exp, priority_ops)
        answer = max(answer, abs(result))
    
    return answer


def calculate_result(exp, priority_ops):
    exp_q = deque(exp)
    mid_q = deque()
    
    for priority_op in priority_ops:
        while exp_q:
            x = exp_q.popleft()
            
            if x.isdigit() or x != priority_op:
                mid_q.append(x)
                continue
            
            n1 = mid_q.pop()
            n2 = exp_q.popleft()
            mid_q.append(calculate(n1, n2, x))
        
        exp_q = mid_q.copy()
        mid_q = deque()
        
    result = exp_q.popleft()
    return int(result)
        

def calculate(number1, number2, op):
    number1, number2 = int(number1), int(number2)
    if op == '+':
        return str(number1 + number2)
    elif op == '-':
        return str(number1 - number2)
    else:
        return str(number1 * number2)
